File: server's lambdas/lego_send_message.py
import json
import boto3
import os

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(os.getenv('CONNECTIONS_TABLE'))

def lambda_handler(event, context):
    print("Event:", event)
    body = json.loads(event['body'])

    user_id = body['userId']
    start_auto_color = body['startAutoColor']
    seconds_offered = body['secondsOffered']
    part1_time = body['part1Time']
    print("before try")
    try:
        print("before update_table_with_seconds_offered")
        update_table_with_seconds_offered(user_id, start_auto_color, seconds_offered, part1_time)
        print("after update_table_with_seconds_offered")
        my_connection_info = get_my_pair_info(user_id, start_auto_color)
        print("after my_connection_info")
        pair_info = get_pair_info(my_connection_info['myPair'], my_connection_info['pairAutoColor'])
        print("before if not pair_info")
        if not pair_info['pairSecondsOffered']:
            return respond_with_message('pair is still playing', 'wait')
        print("after if not pair_info")
        do_deal = calculate_deal(start_auto_color, seconds_offered, pair_info['pairSecondsOffered'])
        print("before update_deal_done")
        update_deal_done(user_id, start_auto_color, do_deal)
        print("after update_deal_done")
        #if pair_info['pairLive'] and pair_info['pairSecondsOffered']: ################
        if pair_info['pairSecondsOffered']:
            print("in if pairSecondsOffered")
            update_deal_done(my_connection_info['myPair'], my_connection_info['pairAutoColor'], do_deal)

        response_body = create_response_body(do_deal, pair_info['pairSecondsOffered'], seconds_offered)
        print("after create_response_body")
        #send_message_to_clients(my_connection_info['myConnectionId'], pair_info['pairConnectionId'], pair_info['pairLive'], response_body)#############
        send_message_to_clients(my_connection_info['myConnectionId'], pair_info['pairConnectionId'], False, response_body)
        print("after send_message_to_clients")
        return {
            'statusCode': 200,
            'body': json.dumps({'message': 'Message sent successfully'})
        }
    except Exception as error:
        print('Exception Error:', error)
        return {
            'statusCode': 500,
            'body': json.dumps({'message': 'Failed to process request'})
        }

def update_table_with_seconds_offered(user_id, start_auto_color, seconds_offered, part1_time):
    params = {
        'TableName': os.getenv('CONNECTIONS_TABLE'),
        'Key': {'ID': user_id, 'startAutoColor': start_auto_color},
        'UpdateExpression': 'set secondsOffered = :seconds, part1Time = :part1Time',
        'ExpressionAttributeValues': {':seconds': seconds_offered, ':part1Time': part1_time},
        'ReturnValues': 'ALL_NEW',
    }
    print("update_table_with_seconds_offered1")
    table.update_item(**params)
    print("update_table_with_seconds_offered2")

def get_my_pair_info(user_id, start_auto_color):
    params = {
        'TableName': os.getenv('CONNECTIONS_TABLE'),
        'Key': {'ID': user_id, 'startAutoColor': start_auto_color},
    }
    response = table.get_item(**params)
    if 'Item' not in response:
        raise Exception('Item not found')

    my_connection_id = response['Item']['connectionId']
    my_pair = response['Item']['myPairId']
    pair_auto_color = 'NO' if start_auto_color == 'YES' else 'YES'
    return {'myConnectionId': my_connection_id, 'myPair': my_pair, 'pairAutoColor': pair_auto_color}

def get_pair_info(my_pair, pair_auto_color):
    params = {
        'TableName': os.getenv('CONNECTIONS_TABLE'),
        'Key': {'ID': my_pair, 'startAutoColor': pair_auto_color},
        'ConsistentRead': True
    }
    response = table.get_item(**params)
    print("response", response)
    #todo: if not such record other not started
    #    if no secondsOffered in response other not finish yet (and maybe not connected any more?)
    if 'Item' not in response:
        raise Exception('Item_pair not found')

    pair_connection_id = response['Item']['connectionId']
    pair_seconds_offered = response['Item']['secondsOffered']
    pair_live = True #is_connection_live(pair_connection_id) ###########
    if not pair_live:
        raise Exception('pair is not connected anymore')

    return {'pairConnectionId': pair_connection_id, 'pairSecondsOffered': pair_seconds_offered, 'pairLive': pair_live}

def calculate_deal(start_auto_color, seconds_offered, pair_seconds_offered):
    if start_auto_color == 'YES':
        return int(seconds_offered) <= int(pair_seconds_offered)
    else:
        return int(seconds_offered) >= int(pair_seconds_offered)

def update_deal_done(user_id, start_auto_color, do_deal):
    params = {
        'TableName': os.getenv('CONNECTIONS_TABLE'),
        'Key': {'ID': user_id, 'startAutoColor': start_auto_color},
        'UpdateExpression': 'set dealDone = :deal',
        'ExpressionAttributeValues': {':deal': do_deal},
        'ReturnValues': 'ALL_NEW',
    }
    table.update_item(**params)

def create_response_body(do_deal, pair_seconds_offered, my_seconds_offered):
    return {
        'message': 'deal results are ' + str(do_deal),
        'ansStatus': 'continue',
        'isdealDone': do_deal,
        'pairSecondsOffered': pair_seconds_offered,
        'mySecondsOffered': my_seconds_offered
    }

def send_message_to_clients(my_connection_id, pair_connection_id, pair_live, response_body):##############
    apigatewaymanagementapi = boto3.client('apigatewaymanagementapi', endpoint_url=os.getenv('API_GATEWAY_ENDPOINT'))
    print("apigatewaymanagementapi", apigatewaymanagementapi)
    print("response_body", response_body)
    print("my_connection_id", my_connection_id)
    if my_connection_id:
        print("in my_connection_id: ", my_connection_id )
        apigatewaymanagementapi.post_to_connection(
            ConnectionId=my_connection_id,
            Data=json.dumps(response_body)
        )
    if pair_live and pair_connection_id:
        print("in pair_connection_id")
        apigatewaymanagementapi.post_to_connection(
            ConnectionId=pair_connection_id,
            Data=json.dumps(response_body)
        )

def send_to_one_client(client_connection_id, response_body, ):
    apigatewaymanagementapi = boto3.client('apigatewaymanagementapi', endpoint_url=os.getenv('API_GATEWAY_ENDPOINT'))
    print("in send_to_one_client: ", client_connection_id )
    apigatewaymanagementapi.post_to_connection(
        ConnectionId=client_connection_id,
        Data=json.dumps(response_body)
    )


def respond_with_message(message, status):
    return {
        'statusCode': 200,
        'body': json.dumps({'message': message, 'ansStatus': status}),
    }

################
def is_connection_live(connection_id):
    apigatewaymanagementapi = boto3.client('apigatewaymanagementapi', endpoint_url=os.getenv('API_GATEWAY_ENDPOINT'))

    try:
        # Attempt to send a ping message to the connection
        apigatewaymanagementapi.post_to_connection(
            ConnectionId=connection_id,
            Data=json.dumps({ 'type': 'ping' })  # Sending a simple "ping" message
        )

        # If no error, the connection is live
        return True
    except apigatewaymanagementapi.exceptions.GoneException:
        # 410 Gone indicates the connection is no longer present
        print(f'Connection {connection_id} is no longer live. (server ERROR will be seen on client and its not a problem ).')
        return False
    except Exception as err:
        # Handle other potential errors
        print(f'Failed to send ping to connection {connection_id}:', err)
        raise err

File: server's lambdas/lego_send_message.mjs
///we will change it to get the user secondsOffered, save it to the user record and to the couple's one
///check if other user already answer. IF YES:
///evaluate if deal done and write to the relevent records
///send message to the other user ( tell him if the deal done and maybe the offered seconds)
///send it to the user, too

const AWS = require('aws-sdk');
const dynamoDb = new AWS.DynamoDB.DocumentClient();
const apigatewaymanagementapi = new AWS.ApiGatewayManagementApi({
    endpoint: process.env.API_GATEWAY_ENDPOINT
});
const CONNECTIONS_TABLE = process.env.TABLE_NAME;

exports.handler = async (event) => {
    console.log("send message");
    const body = JSON.parse(event.body);

    const userId = body.userId;
    const startAutoColor = body.startAutoColor;
    const secondsOffered = body.secondsOffered;
    const part1Time = body.part1Time;

    try {
        ///update my record in the database with my offer and mypart1time
        await updateTableWithSecondsOffered(userId, startAutoColor, secondsOffered, part1Time);
        ///get my record to use myConnectionId later and ID of myPair to allowreding its record 
        const { myConnectionId, myPair, pairAutoColor } = await getMyPairInfo(userId, startAutoColor);
        ///get the record of my pair to use know if he live (connected) and his offer and connectionId from DB
        const { pairConnectionId, pairSecondsOffered, pairLive } = await getPairInfo(myPair, pairAutoColor);

        ///this not send message to client and exit the lambda (we will stay with default message to wait that client created  )
        ///(if pair is not connected we act in the same wa. he need to reconnect and end part1 - we will handle it elsewhere
        if (!pairSecondsOffered) {
            return respondWithMessage('pair is still playing', 'wait');
        }
       
        if (!pairLive) { 
             ////we know now that pairSecondsOffered exsists so pair disconnected after he end patr 1
             ///we could not send him our answer and he must reconnect using 9, when he will do it we will get message so we dont do anithing now and wait
             ///TODO: we can send to me message that game will continue after othe uset will restart using 9
             ///anyway we don't continue
             return;
        }

        ///we have now the pair connected after he end part1
        const doDeal = calculateDeal(startAutoColor, secondsOffered, pairSecondsOffered);
        await updateDealDone(userId, startAutoColor, doDeal);///update deal result in MY record on table lego_couple
        await updateDealDone(myPair, pairAutoColor, doDeal);///update deal result in PAIR record on table lego_couple
        
        ///send deal results with continue to both 
        const theBody = createResponseBody(doDeal, pairSecondsOffered, secondsOffered);
        console.log("before sendMessageToClients")
        await sendMessageToClients(myConnectionId, pairConnectionId, pairLive, theBody);

        return { statusCode: 200, body: JSON.stringify({ message: 'Message sent successfully' }) };
    } catch (error) {
        console.error('Error:', error);
        return { statusCode: 500, body: JSON.stringify({ message: 'Failed to process request' }) };
    }
};

async function updateTableWithSecondsOffered(userId, startAutoColor, secondsOffered, part1Time) {
    const params = {
        TableName: CONNECTIONS_TABLE,
        Key: { ID: userId, startAutoColor: startAutoColor },
        UpdateExpression: "set secondsOffered = :seconds, part1Time = :psrtTime",
        ExpressionAttributeValues: { ":seconds": secondsOffered, ":psrtTime": part1Time },
        ReturnValues: "ALL_NEW",
    };
    await dynamoDb.update(params).promise();
}

async function getMyPairInfo(userId, startAutoColor) {
    const params = {
        TableName: CONNECTIONS_TABLE,
        Key: { ID: userId, startAutoColor: startAutoColor },
    };
    const response = await dynamoDb.get(params).promise();
    if (!response.Item) throw new Error('Item not found');

    const myConnectionId = response.Item.connectionId;
    const myPair = response.Item.myPairId;
    const pairAutoColor = startAutoColor === "YES" ? "NO" : "YES";
    return { myConnectionId, myPair, pairAutoColor };
}

async function getPairInfo(myPair, pairAutoColor) {
    ///read from data base the record of pair
    const params = {
        TableName: CONNECTIONS_TABLE,
        Key: { ID: myPair, startAutoColor: pairAutoColor },
    };
    const response = await dynamoDb.get(params).promise();
    if (!response.Item) throw new Error('Item_pair not found');

    const pairConnectionId = response.Item.connectionId;
    const pairSecondsOffered = response.Item.secondsOffered;
    const pairLive = await isConnectionLive(pairConnectionId);
    if (!pairLive) throw new Error('pair is not connected anymore');

    return { pairConnectionId, pairSecondsOffered, pairLive };
}

function calculateDeal(startAutoColor, secondsOffered, pairSecondsOffered) {
    if (startAutoColor === "YES") {
        return parseInt(secondsOffered) <= parseInt(pairSecondsOffered);
    } else {
        return parseInt(secondsOffered) >= parseInt(pairSecondsOffered);
    }
}

async function updateDealDone(userId, startAutoColor, doDeal) {
    const params = {
        TableName: CONNECTIONS_TABLE,
        Key: { ID: userId, startAutoColor: startAutoColor },
        UpdateExpression: "set dealDone = :deal",
        ExpressionAttributeValues: { ":deal": doDeal },
        ReturnValues: "ALL_NEW",
    };
    await dynamoDb.update(params).promise();
}

function createResponseBody(doDeal, pairSecondsOffered, mySecondsOffered) {
    return {
        message: 'deal results are ' + doDeal,
        ansStatus: 'continue',
        isdealDone: doDeal,
        pairSecondsOffered: pairSecondsOffered,
        mySecondsOffered: mySecondsOffered
    };
}

async function sendMessageToClients(myConnectionId, pairConnectionId, pairLive, theBody) {
    if (myConnectionId) {
        console.log("sendMessageToClients me: " + myConnectionId);
        await apigatewaymanagementapi.postToConnection({
            ConnectionId: myConnectionId,
            Data: JSON.stringify(theBody)
        }).promise();
    }
    if (pairLive && pairConnectionId) {
        console.log("sendMessageToClients pair: " + pairConnectionId)
        await apigatewaymanagementapi.postToConnection({
            ConnectionId: pairConnectionId,
            Data: JSON.stringify(theBody)
        }).promise();
    }
}

function respondWithMessage(message, status) {
    return {
        statusCode: 200,
        body: JSON.stringify({ message: message, ansStatus: status }),
    };
}

async function isConnectionLive(connectionId) {
    try {
        // Attempt to send a ping message to the connection
        await apigatewaymanagementapi.postToConnection({
            ConnectionId: connectionId,
            Data: JSON.stringify({ type: 'ping' }) // Sending a simple "ping" message
        }).promise();

        // If no error, the connection is live
        return true;
    } catch (err) {
        // If an error occurs, it usually means the connection is closed
        if (err.statusCode === 410) { // 410 Gone indicates the connection is no longer present
            console.log(`Connection ${connectionId} is no longer live. (server ERROR will be seen on client and its not a problem ).`);
            return false;
        }
        // Handle other potential errors
        console.error(`Failed to send ping to connection ${connectionId}:`, err);
        throw err;
    }
}

